File: prototype/wasp/packages/wasmvm/wasmlib/go/wasmlib/wasmtypes/scint32.go
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

package wasmtypes

import (
	"encoding/binary"
	"strconv"
)

const ScInt32Length = 4

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

func Int32Decode(dec *WasmDecoder) int32 {
	return Int32FromBytes(dec.FixedBytes(ScInt32Length))
}

func Int32Encode(enc *WasmEncoder, value int32) {
	enc.FixedBytes(Int32ToBytes(value), ScInt32Length)
}

func Int32FromBytes(buf []byte) int32 {
	if len(buf) == 0 {
		return 0
	}
	if len(buf) != ScInt32Length {
		panic("invalid Int32 length")
	}
	return int32(binary.LittleEndian.Uint32(buf))
}

func Int32ToBytes(value int32) []byte {
	tmp := make([]byte, ScInt32Length)
	binary.LittleEndian.PutUint32(tmp, uint32(value))
	return tmp
}

func Int32FromString(value string) int32 {
	return int32(IntFromString(value, 32))
}

func Int32ToString(value int32) string {
	return strconv.FormatInt(int64(value), 10)
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

type ScImmutableInt32 struct {
	proxy Proxy
}

func NewScImmutableInt32(proxy Proxy) ScImmutableInt32 {
	return ScImmutableInt32{proxy: proxy}
}

func (o ScImmutableInt32) Exists() bool {
	return o.proxy.Exists()
}

func (o ScImmutableInt32) String() string {
	return Int32ToString(o.Value())
}

func (o ScImmutableInt32) Value() int32 {
	return Int32FromBytes(o.proxy.Get())
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

type ScMutableInt32 struct {
	ScImmutableInt32
}

func NewScMutableInt32(proxy Proxy) ScMutableInt32 {
	return ScMutableInt32{ScImmutableInt32{proxy: proxy}}
}

func (o ScMutableInt32) Delete() {
	o.proxy.Delete()
}

func (o ScMutableInt32) SetValue(value int32) {
	o.proxy.Set(Int32ToBytes(value))
}

File: prototype/wasp/packages/wasmvm/wasmlib/src/wasmtypes/scint32.rs
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

use crate::*;

pub const SC_INT32_LENGTH: usize = 4;

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

pub fn int32_decode(dec: &mut WasmDecoder) -> i32 {
    int32_from_bytes(&dec.fixed_bytes(SC_INT32_LENGTH))
}

pub fn int32_encode(enc: &mut WasmEncoder, value: i32) {
    enc.fixed_bytes(&int32_to_bytes(value), SC_INT32_LENGTH);
}

pub fn int32_from_bytes(buf: &[u8]) -> i32 {
    let len = buf.len();
    if len == 0 {
        return 0;
    }
    if len != SC_INT32_LENGTH {
        panic("invalid Int32 length");
    }
    i32::from_le_bytes(buf.try_into().expect("WTF?"))
}

pub fn int32_to_bytes(value: i32) -> Vec<u8> {
    value.to_le_bytes().to_vec()
}

pub fn int32_from_string(value: &str) -> i32 {
    value.parse::<i32>().unwrap()
}

pub fn int32_to_string(value: i32) -> String {
    value.to_string()
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

pub struct ScImmutableInt32 {
    proxy: Proxy,
}

impl ScImmutableInt32 {
    pub fn new(proxy: Proxy) -> ScImmutableInt32 {
        ScImmutableInt32 { proxy }
    }

    pub fn exists(&self) -> bool {
        self.proxy.exists()
    }

    pub fn to_string(&self) -> String {
        int32_to_string(self.value())
    }

    pub fn value(&self) -> i32 {
        int32_from_bytes(&self.proxy.get())
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

// value proxy for mutable i32 in host container
pub struct ScMutableInt32 {
    proxy: Proxy,
}

impl ScMutableInt32 {
    pub fn new(proxy: Proxy) -> ScMutableInt32 {
        ScMutableInt32 { proxy }
    }

    pub fn delete(&self) {
        self.proxy.delete();
    }

    pub fn exists(&self) -> bool {
        self.proxy.exists()
    }

    pub fn set_value(&self, value: i32) {
        self.proxy.set(&int32_to_bytes(value));
    }

    pub fn to_string(&self) -> String {
        int32_to_string(self.value())
    }

    pub fn value(&self) -> i32 {
        int32_from_bytes(&self.proxy.get())
    }
}
